source/Variant.cpp

```diff
@@ -0,0 +1,255 @@
+/* Variant.cpp
+Copyright (c) 2019 by Jonathan Steck
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+*/
+
+#include "Variant.h"
+
+#include "DataNode.h"
+#include "GameData.h"
+#include "Random.h"
+#include "Ship.h"
+#include "WeightedVariant.h"
+
+#include <algorithm>
+#include <iterator>
+
+using namespace std;
+
+
+
+// Construct and Load() at the same time.
+Variant::Variant(const DataNode &node)
+{
+	Load(node);
+}
+
+
+
+void Variant::Load(const DataNode &node)
+{
+	// If this variant is being loaded with a second token that is not a number,
+	// then it must be being loaded from GameData, and therefore must have its
+	// name saved.
+	if(node.Size() >= 2 && !node.IsNumber(1))
+		name = node.Token(1);
+	
+	// If Load() has already been called once on this variant, any subsequent
+	// calls will replace the contents instead of adding to them.
+	bool reset = !variants.empty() || !ships.empty();
+	
+	for(const DataNode &child : node)
+	{
+		bool add = (child.Token(0) == "add");
+		bool remove = (child.Token(0) == "remove");
+		if((add || remove) && child.Size() == 1)
+		{	
+			child.PrintTrace("Skipping invalid \"" + child.Token(0) + "\" tag:");
+			continue;
+		}
+		bool variant = (child.Token(add || remove) == "variant");
+		
+		if(remove)
+		{
+			if(variant)
+			{
+				// If given a full definition of one of this fleet's variant members, remove the variant.
+				Variant toRemove(child);
+				auto VariantToRemove = [&](const WeightedVariant &v) noexcept -> bool
+				{
+					return v.Get() == toRemove;
+				};
+				
+				auto removeIt = find_if(variants.begin(), variants.end(), VariantToRemove);
+				if(removeIt != variants.end())
+					variants.erase(remove_if(removeIt, variants.end(), VariantToRemove), variants.end());
```

Comment:
You don't need the `find_if`. You can store the iterator returned by `remove_if` to check whether any elements were removed. Same thing below.